### ISSUE TITLE
auto.tfvars should be merged like prior to 0.11

### ISSUE THREAD

**Issue by schollii** [hashicorp/terraform#28899]
### Current Terraform Version

```
0.15.3
```

### Use-cases

You want to separate the configuration values between several files, and some of these values are in maps. Eg 
```
// terraform.tfvars
portal = {
   something = {
      param1 = "something"
  }
}

// other.tfvars
portal = {
   something = {
      param2 = "something"
  }
}
```

### Attempted Solutions

There is no merge so the only other solution would be to programmatically generate the tfvars file so the merge can happen in code (say via python, Go). 

### Proposal

Restore the merge behavior from before terraform 0.12.

**Comment by apparentlymart** [comment]
Hi @schollii! Thanks for sharing this use-case.

As you say, we currently have no direct answer for splitting a particular input variable across multiple `.tfvars` files. The behavior you refer to in v0.11 was a controversial one because it meant that there was no single location that could be considered the definition of the input variable, which led to two perceived problems:

* For human readers, the only way to understand what a particular input variable is set to would be to consult _all_ of the files and mentally compute the merger of all of them. Some users would write a new `.tfvars` file to override a particular variable and then be confused as to why the result ended up being a mixture of the new and old values.
* For Terraform itself, there was no single location that Terraform could consider to be the definition of the input variable for the purpose of reporting errors. Prior to v0.11 that didn't really matter because those Terraform versions had generally poor error messages with no source location tracking, but an important goal for Terraform v0.12 was for Terraform to be able to trace back every value to a specific definition location as part of the effort to return actionable error messages.

With that said, I don't think restoring the Terraform v0.11 behavior exactly would be desirable, though even if it were we are also blocked from doing so by backward-compatibility, because this would subtly change the behavior of existing configurations.

Given that, I think we'll need to use this issue to discuss the higher level goal, rather than the specific solution you proposed. In particular, I think there are two things to decide here:
* First, we need to think about the tradeoff of splitting a definition across multiple files: is it a good idea to allow that, given our experiences with the behavior in v0.11 and earlier? I understand that you have some use-cases for it, and it would be helpful if you could share some more detail about those so that we can think about how they interact with the conflicting concern of having a single definition point for each variable.
* If we _do_ ultimately conclude that it's desirable to have such a feature, we'll need to design a new way to achieve it that's backwards compatible with the Terraform v1.0 behavior. Ideally such a design would aim to achieve a compromise which avoids reintroducing the disadvantages of the v0.11 design which I described above, such as using an _explicit_ syntax to gather variable values from another file, rather than it being implicit. (The compatibility constraints on the `.tfvars` file format are quite severe though, because the `.tfvars.json` variant is designed to take JSON values exactly as written and not interpret them as Terraform template syntax at all, for ease of programmatically generating JSON values.)

I will also note that I'd typically advise against using input variables for such complex structures. Input variables are more typically for values that need to vary from one run to the next. For complex values that describe _what to manage_ I'd usually suggest encoding that directly in the main configuration, where you have the benefit of using the full Terraform language as part of the definition. However, I'm still interested to see your use-cases since I assume you've considered those possibilities and concluded that root module input variables are the best answer for your particular scenario.

Thanks again!

**Comment by schollii** [comment]
I could see how the original behavior could be confusing, so just going to back to pre-0.12 is not a good solution. My use case is that I have a "main" terraform.tfvars.json that gets generated programmatically, and a secondary portals.auto.tfvars.json. The main one defines one or more portals, such as their ID, domain name etc, things that are not likely to change ever. The secondary one I'd like to define *only* the IP address whitelist for each portal, that will get used by an ssh security group for  that portal. This file could change regularly as people get new WAN IP addresses. So the two files ideally would look like this: 
```
// terraform.tfvars
portals = {
   id1 = {
      name = "name1"
      domain = "domain1.com"
  }
  id2 = {
      name = "name2"
      domain = "domain2.com"
  }
}

// other.tfvars
portals = {
   id1 = {
      cidr_whitelist = {
        "ipA/32" = "Joe Office"
        "ipB/32" = "Jane Office"
        "ipC/30" = "HQ"
     }
  }
  id2 = {
        "ipD/32" = "Carolyn Office"
  }
}
```

This does not work now because the auto tfvars overwrites the id1 and id2 from terraform.tfvars, rather than getting merged. 

I would welcome an explicit inclusion mechanism. The terraform.tfvars could have "merge" statements. I would firstly allow them at the end since then it is clear what the "base is" (whatever comes before). Could also allow that beginning but then what comes after the beginning should get merged into those. Eg
```
// terraform.tfvars
merge a.tfvars // everything in a loaded first
// then merge the following: 
a = {b = "c", d="e"}
// and finally merge c:
merge c.tfvars
```
Or perhaps even 
```
// terraform.tfvars
a = {
  b = "c"
  include d.tfvars
}
// d.tfvars
d = "e"
```
Either way, it should be possible to get a "merged inputs" with or without sources, eg `terraform inputs` would show the merged inputs, and adding the `--show-sources` would show the source of each value:
```
$ terraform inputs --show-sources
sources:
#1: terraform.tfvars
#2: a.tfvars
#3: ../b/b.tfvars

merged:
a [#1] = {
  b [#1, #2] = "c"
  d [#2] = "e"
}
```

**Comment by apparentlymart** [comment]
Thanks for sharing that use-case, @schollii! We'll consider it.

In the mean time, my suggestion to get this done within the current Terraform design would be to split this across two variables: one for the part that changes rarely and another for the part that changes frequently.

```hcl
variable "portals" {
  type = map(object({
    name   = string
    domain = string
  }))
}

variable "portal_cidr_whitelist" {
  type = map(map(string))
}

locals {
  portals = {
    for k, portal in var.portals : k => merge(
      portal,
      { cidr_whitelist = var.portal_cidr_whitelist[k] },
    ),
  }
}
```

```hcl
portals = {
   id1 = {
      name = "name1"
      domain = "domain1.com"
  }
  id2 = {
      name = "name2"
      domain = "domain2.com"
  }
}
```

```hcl
portal_cidr_whitelist = {
  id1 = {
    "ipA/32" = "Joe Office"
    "ipB/32" = "Jane Office"
    "ipC/30" = "HQ"
  }
  id2 = {
    "ipD/32" = "Carolyn Office"
  }
}
```

Then in the rest of your module, use `local.portals` instead of `var.portals` to get the version where the two data structures are already aggregated together.

**Comment by schollii** [comment]
Yeah I thought of same. It is not better (or worse), however, than moving the portals element to a separate file. Both workarounds are ok, but neither is as nice as a proper (albeit explicit) merge.
